File: Examples/ObjC/Headless/SlyceSDK-Headless-Demo-ObjC/TextViewViewController.m

#import "TextViewViewController.h"


@interface TextViewViewController ()

@property (nonatomic, strong) UITextView *textView;

@end

@implementation TextViewViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.textView = [[UITextView alloc] init];
    [self.textView setEditable:NO];
    [self.textView setScrollEnabled:YES];
    self.textView.showsHorizontalScrollIndicator = NO;
    self.textView.alwaysBounceHorizontal = NO;
    self.textView.alwaysBounceVertical = YES;
    [self.textView setDirectionalLockEnabled:YES];
    self.textView.font = [UIFont systemFontOfSize:20];
    self.textView.textColor = [UIColor blackColor];
    self.textView.backgroundColor = [UIColor whiteColor];
    self.textView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.textView];
    
    [[self.textView.topAnchor constraintEqualToAnchor:self.view.topAnchor] setActive:YES];
    [[self.textView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor] setActive:YES];
    [[self.textView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor] setActive:YES];
    [[self.textView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor] setActive:YES];
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    return UIInterfaceOrientationMaskPortrait;
}

- (void)appendTaskToTextView:(SlyceSearchTask *)task finished:(BOOL)finished {
    NSString *stateString = [NSString stringWithFormat:@"\nSlyceSearchTask %@:\n\t", finished ? @"finished" : @"created"];
    NSMutableAttributedString *string = [[NSMutableAttributedString alloc] initWithString:stateString attributes:@{NSFontAttributeName : [UIFont systemFontOfSize:18 weight:UIFontWeightBold]}];
    
    [self appendAttributedStringToTextView:string];
    [self appendImageToTextView:task.request.image];
}

- (void)appendErrorsToTextView:(NSArray<NSError*>*)errors {
    NSMutableAttributedString *string = [[NSMutableAttributedString alloc] initWithString:@"\nSlyceSearchTask errors:\n\t" attributes:@{NSFontAttributeName : [UIFont systemFontOfSize:18 weight:UIFontWeightBold]}];
    
    for (NSError *error in errors) {
        [string appendAttributedString:[[NSAttributedString alloc] initWithString:[error.localizedDescription stringByAppendingString:@"\n"]]];
    }
    
    [self appendAttributedStringToTextView:string];
}

- (void)appendUpdateToTextView:(SlyceSearchResponseUpdate *)update {
    NSMutableAttributedString *string = [[NSMutableAttributedString alloc] initWithString:@"\nSlyceSearchTask update:\n\t" attributes:@{NSFontAttributeName : [UIFont systemFontOfSize:18 weight:UIFontWeightBold]}];
    
    [string appendAttributedString:[[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"Type: %li", update.type]]];
    [string appendAttributedString:[[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"\n\tValue: %@", update.value]]];
    
    [self appendAttributedStringToTextView:string];
}


- (void)appendResponseToTextView:(SlyceSearchResponse *) response {
    NSMutableAttributedString *string = [[NSMutableAttributedString alloc] initWithString:@"\nSlyceSearchTask response:\n\t" attributes:@{NSFontAttributeName : [UIFont systemFontOfSize:18 weight:UIFontWeightBold]}];
    
    [string appendAttributedString:[[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"Results: %@", response.results.debugDescription]]];
    [string appendAttributedString:[[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"\n\tTags: %@", response.tags.debugDescription]]];
    
    [self appendAttributedStringToTextView:string];
}


- (void)appendStringToTextView:(NSString *)string {
    [self appendAttributedStringToTextView:[[NSAttributedString alloc] initWithString:string]];
}


- (void)appendImageToTextView:(UIImage *)image {
    NSTextAttachment *imageAttachment = [[NSTextAttachment alloc] init];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        CGFloat maxSize = self.view.frame.size.width - 50.0;
        
        if (image.size.height > maxSize || image.size.height > maxSize) {
            CGFloat newWidth = maxSize;
            CGFloat scale = newWidth / image.size.width;
            CGFloat newHeight = image.size.height * scale;
            UIGraphicsBeginImageContext(CGSizeMake(newWidth, newHeight));
            [image drawInRect:CGRectMake(0, 0, newWidth, newHeight)];
            UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
            UIGraphicsEndImageContext();
            imageAttachment.image = newImage;
        } else {
            imageAttachment.image = image;
        }
    });
    
    
    [self appendAttributedStringToTextView:[NSAttributedString attributedStringWithAttachment:imageAttachment]];
}

- (void)appendAttributedStringToTextView:(NSAttributedString *)attributedString {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSMutableAttributedString *appendingString = [[NSMutableAttributedString alloc] initWithAttributedString:self.textView.attributedText];
        [appendingString appendAttributedString:attributedString];
        
        NSRange endOfTextRange = NSMakeRange(appendingString.string.length - 1, 1);
        self.textView.attributedText = appendingString;
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.textView scrollRangeToVisible:endOfTextRange];
        });
    });
}

@end

File: Examples/Swift/Headless/SlyceSDK-Headless-Demo-Swift/TextViewViewController.swift
import UIKit
import SlyceSDK

class TextViewViewController: UIViewController
{
    let textView = UITextView()
    
    override func viewDidLoad()
    {
        super.viewDidLoad()

        // This view controller shows a UITextView, and implements a few
        // convenience methods that make it easy to add images and text
        // to the text view, and scroll to the bottom on each update.
        //
        // None of this code is necessary to run Slyce and
        // is only included to help navigate the example app.

        view.backgroundColor = UIColor.white
        
        textView.isEditable = false
        textView.isScrollEnabled = true
        textView.showsHorizontalScrollIndicator = false
        textView.alwaysBounceHorizontal = false
        textView.alwaysBounceVertical = true
        textView.isDirectionalLockEnabled = true
        textView.font = UIFont.systemFont(ofSize: 20)
        textView.textColor = UIColor.black
        textView.backgroundColor = UIColor.white
        textView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(textView)
        
        textView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        textView.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        textView.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        textView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
    }
    
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask
    {
        return .portrait
    }
    
    func appendTaskToTextView(_ task: SlyceSearchTask, _ finished: Bool = false)
    {
        let stateString = finished ? "finished" : "created"
        let string = NSMutableAttributedString(string: "\nSlyceSearchTask "+stateString+":\n\t", attributes: [NSAttributedStringKey.font : UIFont.systemFont(ofSize: 18, weight: .bold)])
        appendAttributedStringToTextView(string)
        appendImageToTextView(task.request.image!)
    }
    
    func appendErrorsToTextView(_ errors: [Error])
    {
        let string = NSMutableAttributedString(string: "\nSlyceSearchTask errors:\n\t", attributes: [NSAttributedStringKey.font : UIFont.systemFont(ofSize: 18, weight: .bold)])
        
        for error in errors {
            string.append(NSAttributedString(string: error.localizedDescription.appending("\n")))
        }
        
        appendAttributedStringToTextView(string)
    }
    
    func appendUpdateToTextView(_ update: SlyceSearchResponseUpdate)
    {
        let string = NSMutableAttributedString(string: "\nSlyceSearchTask update:\n\t", attributes: [NSAttributedStringKey.font : UIFont.systemFont(ofSize: 18, weight: .bold)])
        
        string.append(NSAttributedString(string: "Type: ".appending(NSNumber(value: Int8(update.type.rawValue)).stringValue)))
        string.append(NSAttributedString(string: "\n\tValue: ".appending(update.value.debugDescription)))

        appendAttributedStringToTextView(string)
    }
    
    func appendResponseToTextView(_ response: SlyceSearchResponse)
    {
        let string = NSMutableAttributedString(string: "\nSlyceSearchTask response:\n\t", attributes: [NSAttributedStringKey.font : UIFont.systemFont(ofSize: 18, weight: .bold)])
        
        string.append(NSAttributedString(string: "Results: ".appending(response.results.debugDescription)))
        string.append(NSAttributedString(string: "\n\tTags: ".appending(response.tags.debugDescription)))
        
        appendAttributedStringToTextView(string)
    }
    
    func appendStringToTextView(_ string: String)
    {
        appendAttributedStringToTextView(NSAttributedString(string: string))
    }
    
    func appendImageToTextView(_ image: UIImage)
    {
        let imageAttachment = NSTextAttachment()

        OperationQueue.main.addOperation {
            // Resize image if too large to show elegantly in text view.
            let maxSize: CGFloat = self.view.frame.size.width - 50.0
            if image.size.height > maxSize || image.size.width > maxSize {
                let newWidth = maxSize
                let scale = newWidth / image.size.width
                let newHeight = image.size.height * scale
                UIGraphicsBeginImageContext(CGSize(width: newWidth, height: newHeight))
                image.draw(in: CGRect(x: 0, y: 0, width: newWidth, height: newHeight))
                let newImage = UIGraphicsGetImageFromCurrentImageContext()
                UIGraphicsEndImageContext()
                imageAttachment.image = newImage
            } else {
                imageAttachment.image = image
            }
        }

        appendAttributedStringToTextView(NSAttributedString(attachment: imageAttachment))
    }
    
    func appendAttributedStringToTextView(_ attributedString: NSAttributedString)
    {
        OperationQueue.main.addOperation {
            let appendingString = NSMutableAttributedString(attributedString: self.textView.attributedText)
            appendingString.append(attributedString)
            
            let endOfTextRange = NSRange(location: appendingString.string.count-1, length: 1)
            
            self.textView.attributedText = appendingString
            
            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.25, execute: {
                self.textView.scrollRangeToVisible(endOfTextRange)
            })
        }
    }
}
